controller/src/main/java/io/pravega/controller/server/eventProcessor/LocalController.java

```diff
@@ -201,11 +215,75 @@ public LocalController(ControllerService controller, boolean authorizationEnable
                     return true;
                 default:
                     throw new ControllerFailureException("Unknown return status creating ReaderGroup " + scopedRGName
-                            + " " + x.getStatus());
+                            + " " + x);
             }
         });
     }
 
+    CompletableFuture<CreateReaderGroupStatus.Status> createReaderGroupInternal(String scope, String rgName, ReaderGroupConfig config, final long createTimestamp) {
+        Preconditions.checkNotNull(scope, "ReaderGroup scope is null");
+        Preconditions.checkNotNull(rgName, "ReaderGroup name is null");
+        Preconditions.checkNotNull(config, "ReaderGroup config is null");
+        Preconditions.checkArgument(createTimestamp >= 0);
+        try {
+            NameUtils.validateReaderGroupName(rgName);
+        } catch (IllegalArgumentException | NullPointerException e) {
+            log.warn("Create ReaderGroup failed due to invalid name {}", rgName);
+            return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.INVALID_RG_NAME);
+        }
+        StreamMetadataStore streamMetadataStore = controller.getStreamStore();
+        final RGOperationContext context = streamMetadataStore.createRGContext(scope, rgName);
+        return RetryHelper.withRetriesAsync(() -> {
+            // 1. check if scope with this name exists...
+            return streamMetadataStore.checkScopeExists(scope)
+                    .thenCompose(exists -> {
+                    if (!exists) {
+                            return CompletableFuture.completedFuture(CreateReaderGroupStatus.Status.SCOPE_NOT_FOUND);
+                    }
+                    //2. check state of the ReaderGroup
+                    return Futures.exceptionallyExpecting(streamMetadataStore.getReaderGroupState(scope, rgName, true, null, localExecutor),
+                           e -> Exceptions.unwrap(e) instanceof StoreException.DataNotFoundException, ReaderGroupState.UNKNOWN)
+                           .thenCompose(state -> {
+                           if (state.equals(ReaderGroupState.UNKNOWN) || state.equals(ReaderGroupState.CREATING)) {
+                              return streamMetadataStore.addReaderGroupToScope(scope, rgName, config.getReaderGroupId())
+                              .thenCompose(v -> streamMetadataStore.createReaderGroup(scope, rgName, config, createTimestamp, context, localExecutor)
```

Comment:
i suggest we move all of the code below this to a common helper method in streamMetadataTasks which is invoked both from here and from CreateRGTask

Reply:
done